File: LivrusVirtus/VCPP6/FLTK/LibMySQL/VenderLivros.cpp
#include "Principal.h"
#include "VenderLivros.h"
#include "RotinasGlobais.h"
#include "ConsultasSQL.h"
#include "ConClientes.h"
#include "ConLivros.h"

CfrmVenderLivros *frmVenderLivros;
LISTASTR frmVenderLivros_slPrecos, frmVenderLivros_slISBNs, 
  frmVenderLivros_slQtdEstoque;

void frmVenderLivros_btnPClienteClick_cb(Fl_Button* b, void* v);
void frmVenderLivros_btnAdLivrosClick_cb(Fl_Button* b, void* v);
void frmVenderLivros_btnPLivroClick_cb(Fl_Button* b, void* v);
void frmVenderLivros_lstVendasSelect_cb(Fl_Browser* b, void* v);
void frmVenderLivros_btnVenderClick_cb(Fl_Button* b, void* v);
void frmVenderLivros_btnLimparClick_cb(Fl_Button* b, void* v);
void frmVenderLivros_btnFecharClick_cb(Fl_Button* b, void* v);

CfrmVenderLivros::CfrmVenderLivros()
       : Fl_Double_Window(511, 316, "Vender Livros") {
	IniciaControles();
}

CfrmVenderLivros::~CfrmVenderLivros() { }

void CfrmVenderLivros::IniciaControles() {
  this->box(FL_FLAT_BOX);
  this->color(FL_BACKGROUND_COLOR);
  this->selection_color(FL_BACKGROUND_COLOR);
  this->labeltype(FL_NO_LABEL);
  this->labelfont(0);
  this->labelsize(14);
  this->labelcolor(FL_BLACK);
  this->user_data((void*)(this));
  this->align(FL_ALIGN_TOP);
  this->when(FL_WHEN_RELEASE);
  edtDtHrVenda = new Fl_Output(10, 20, 160, 25, "Data/Hora Venda:");
  edtDtHrVenda->align(FL_ALIGN_TOP_LEFT);
  edtDtHrVenda->deactivate();

  edtCPF = new Fl_Input(10, 65, 160, 25, "CPF:");
  edtCPF->callback((Fl_Callback*)frmVenderLivros_edtCPFKillFocus_cb);
  edtCPF->align(FL_ALIGN_TOP_LEFT);
  edtCPF->maximum_size(14);

  edtCliente = new Fl_Output(170, 65, 310, 25);
  edtCliente->deactivate();

  btnPCliente = new Fl_Button(480, 65, 25, 25, "...");
  btnPCliente->callback((Fl_Callback*)frmVenderLivros_btnPClienteClick_cb);

  edtISBN = new Fl_Input(10, 110, 160, 25, "ISBN:");
  edtISBN->callback((Fl_Callback*)frmVenderLivros_edtISBNKillFocus_cb);
  edtISBN->align(FL_ALIGN_TOP_LEFT);
  edtISBN->maximum_size(13);

  edtLivro = new Fl_Output(170, 110, 280, 25);
  edtLivro->deactivate();

  btnAdLivros = new Fl_Button(450, 110, 25, 25, "+");
  btnAdLivros->callback((Fl_Callback*)frmVenderLivros_btnAdLivrosClick_cb);

  btnPLivro = new Fl_Button(480, 110, 25, 25, "...");
  btnPLivro->callback((Fl_Callback*)frmVenderLivros_btnPLivroClick_cb);

  lstLivros = new Fl_Browser(10, 155, 495, 95, "Livros para vender:");
  lstLivros->type(1);
  lstLivros->callback((Fl_Callback*)frmVenderLivros_lstVendasSelect_cb);
  lstLivros->align(FL_ALIGN_TOP_LEFT);

  edtPrecoTotal = new Fl_Output(350, 255, 155, 25, "Pre\347o Total:");
  edtPrecoTotal->deactivate();

  btnVender = new Fl_Button(270, 285, 75, 25, "&Vender");
  btnVender->callback((Fl_Callback*)frmVenderLivros_btnVenderClick_cb);

  btnLimpar = new Fl_Button(350, 285, 75, 25, "&Limpar");
  btnLimpar->callback((Fl_Callback*)frmVenderLivros_btnLimparClick_cb);

  btnFechar = new Fl_Button(430, 285, 75, 25, "&Fechar");
  btnFechar->callback((Fl_Callback*)frmVenderLivros_btnFecharClick_cb);

  this->end();
  this->resizable(this);

  edtDtHrVenda->value(Rotinas->retDataHoraStr(true, true));
  nPrecoTotal = 0;
}

void CfrmVenderLivros::LimpaDados() {
  edtDtHrVenda->value(Rotinas->retDataHoraStr(true, true));  
  edtCPF->value("");
  edtCliente->value("");
  edtISBN->value("");
  edtLivro->value("");
  lstLivros->clear();
  nPrecoTotal = 0;
  edtPrecoTotal->value("R$ 0,00");
  edtCPF->take_focus();
}

bool CfrmVenderLivros::ValidaDados() {
  if (Rotinas->ValidaCampo(edtCPF->value(), STR_CPFNINF))
    return false;
  if (lstLivros->size()==0) {
    fl_alert(STR_LIVVENNINF);
    return false;
  }
return true;
}

bool CfrmVenderLivros::SalvaLista()
{ string sCPF, sISBN, sDataHora, sDTHR;
  char sPrecoTotal[TAM_MAX_STR], sQtdEstoque[TAM_MAX_STR];
  int nCont, nQtd;
  int nQtdEstoque;

  nQtd = lstLivros->size();
  sCPF = edtCPF->value();
  sprintf(sPrecoTotal, "%.2f", nPrecoTotal);
  
  sDTHR = edtDtHrVenda->value();
  sDataHora = Rotinas->ConverteDataHoraStr(sDTHR, true, 
    "%m/%d/%Y","%H:%M:%S");
  
  for (nCont=0; nCont < nQtd; nCont++) {
    sISBN = ListaStr->pesquisaItemOrd(frmVenderLivros_slISBNs, nCont);
 
    Rotinas->AtualizaDados(gConexao, ConsSQL->Venda('I',sISBN, sCPF, 
      sDataHora,Rotinas->VirgulaParaPonto(sPrecoTotal,false),""));
  
    nQtdEstoque = 
      atoi(ListaStr->pesquisaItemOrd(frmVenderLivros_slQtdEstoque, nCont));
    if (nQtdEstoque > 0) nQtdEstoque--;
    sprintf(sQtdEstoque, "%d", nQtdEstoque);
    Rotinas->AtualizaDados(gConexao, ConsSQL->Venda('U',sISBN, 
      "","","", sQtdEstoque));
  }
return true;
}

void frmVenderLivros_edtCPFKillFocus_cb(Fl_Input* i, void* v){
string sCliente;
LISTAMSTR ConsCampo;

  if (strlen(frmVenderLivros->edtCPF->value()) > 0) {
    if (Rotinas->validaCPF(frmVenderLivros->edtCPF->value())) {      
      sCliente = Rotinas->ConsultaCampoDesc(&ConsCampo, ConsSQL->Cliente('S',
      frmVenderLivros->edtCPF->value(),"","","","","","","","",""),
      STR_CLINENC); 
      if (sCliente.length()==0) {
        frmVenderLivros->edtCPF->value("");
        frmVenderLivros->edtCPF->take_focus();
      } else
         frmVenderLivros->edtCliente->value(sCliente.c_str());
    } else {
      fl_alert(STR_CPFINV);
      frmVenderLivros->edtCPF->value("");
      frmVenderLivros->edtCPF->take_focus();
    }
  } else
     frmVenderLivros->edtCliente->value("");
}

void frmVenderLivros_btnPClienteClick_cb(Fl_Button* b, void* v){
  frmConClientes = new CfrmConClientes(1);
  frmConClientes->show();
}

void frmVenderLivros_edtISBNKillFocus_cb(Fl_Input* i, void* v){
string sLivro;
int nQtdEstoque;
LISTAMSTR ConsCampo;

  if (strlen(frmVenderLivros->edtISBN->value()) > 0) {
    if (Rotinas->validaISBN(frmVenderLivros->edtISBN->value())) {      
      sLivro = Rotinas->ConsultaCampoDesc(&ConsCampo, ConsSQL->Livro('Q',
      frmVenderLivros->edtISBN->value(),"","","","","","","","",""),
      STR_LIVNENC);
      if (sLivro.length()==0) {
        frmVenderLivros->edtISBN->value("");
      	frmVenderLivros->edtISBN->take_focus();
      } else {
         frmVenderLivros->edtLivro->value(sLivro.c_str());
         nQtdEstoque = atoi(ListaStr->pesquisaItemOrdM(ConsCampo,3,0));
         if (nQtdEstoque > 0) {
           Rotinas_sPreco = ListaStr->pesquisaItemOrdM(ConsCampo,2,0);
           Rotinas_sQtdEstoque = ListaStr->pesquisaItemOrdM(ConsCampo,3,0);
         } else
           fl_alert(STR_LIVNEST);
      }
    } else {
      fl_alert(STR_ISBNINV);
      frmVenderLivros->edtISBN->value("");
    	frmVenderLivros->edtISBN->take_focus();
    }
  } else
     frmVenderLivros->edtLivro->value("");
}

void frmVenderLivros_btnAdLivrosClick_cb(Fl_Button* b, void* v){
double nPreco; 
string sLivros, sISBN, sEdtLivro;
char sPrecoTotal[TAM_MAX_STR];

 sISBN = frmVenderLivros->edtISBN->value();
 sEdtLivro = frmVenderLivros->edtLivro->value();
 if (sISBN.length() > 0) {
   sLivros = sISBN + " - " + sEdtLivro + " - R$ " + Rotinas_sPreco;
   if (Rotinas->VerificaItemLista(frmVenderLivros->lstLivros, sLivros)==false) {
   	 ListaStr->insereFim(&frmVenderLivros_slISBNs, (char*)sISBN.c_str()); 
   	 ListaStr->insereFim(
       &frmVenderLivros_slPrecos, (char*)Rotinas_sPreco.c_str()); 
   	 ListaStr->insereFim(
       &frmVenderLivros_slQtdEstoque, (char*)Rotinas_sQtdEstoque.c_str()); 
     sLivros = "@f" + sLivros;
     frmVenderLivros->lstLivros->add(sLivros.c_str());
     nPreco = atof(Rotinas_sPreco.c_str());
     frmVenderLivros->nPrecoTotal = frmVenderLivros->nPrecoTotal + nPreco;
   }
   frmVenderLivros->edtISBN->value("");
   frmVenderLivros->edtLivro->value("");
   sprintf(sPrecoTotal, "R$ %.2f", frmVenderLivros->nPrecoTotal);
   frmVenderLivros->edtPrecoTotal->value(
     Rotinas->VirgulaParaPonto(sPrecoTotal,true));
 }
}

void frmVenderLivros_btnPLivroClick_cb(Fl_Button* b, void* v){
  frmConLivros = new CfrmConLivros(1);
  frmConLivros->show();
}

void frmVenderLivros_lstVendasSelect_cb(Fl_Browser* b, void* v){
char sPrecoTotal[TAM_MAX_STR];
double nPreco;
int nSelIndice;

  if (Fl::event_clicks()==1) { //duplo-clique
    if (frmVenderLivros->lstLivros->size() > 0) {   
      nSelIndice = frmVenderLivros->lstLivros->value();   
      if (frmVenderLivros->lstLivros->selected(nSelIndice)) {
        nPreco = atof(ListaStr->pesquisaItemOrd(
          frmVenderLivros_slPrecos,nSelIndice));
        frmVenderLivros->nPrecoTotal=frmVenderLivros->nPrecoTotal-nPreco;
        sprintf(sPrecoTotal, "R$ %.2f", frmVenderLivros->nPrecoTotal);
        frmVenderLivros->edtPrecoTotal->value(
          Rotinas->VirgulaParaPonto(sPrecoTotal,true));
   	    ListaStr->removeItemOrd(&frmVenderLivros_slISBNs, nSelIndice, true);
        ListaStr->removeItemOrd(&frmVenderLivros_slPrecos, nSelIndice, true);         
        ListaStr->removeItemOrd(
          &frmVenderLivros_slQtdEstoque, nSelIndice, true);
        frmVenderLivros->lstLivros->remove(nSelIndice);
      }	
    }
  }
}

void frmVenderLivros_btnVenderClick_cb(Fl_Button* b, void* v){
  if (frmVenderLivros->ValidaDados()) {
    if (frmVenderLivros->SalvaLista()) {
      fl_alert(STR_VENRESUC);
      frmVenderLivros->LimpaDados();
    }
  }	
}

void frmVenderLivros_btnLimparClick_cb(Fl_Button* b, void* v){
  frmVenderLivros->LimpaDados();
}

void frmVenderLivros_btnFecharClick_cb(Fl_Button* b, void* v){
  frmVenderLivros->hide();
}

int CfrmVenderLivros::handle(int event) {
  if (event == FL_HIDE) {
    frmPrincipal->BarraDeStatus2->label("");
    frmPrincipal->tlbVenda->activate();
    frmPrincipal->mnuVenVenderLivros->activate();
  }
  return Fl_Double_Window::handle(event);
}

File: LivrusVirtus/FoxToolKit/FXRuby/VenderLivros.rb
require 'fox'
require 'fox/responder'
require 'Cadastros'
require 'Principal'
require 'RotinasGlobais'
require 'ConsultasSQL'
require 'ConClientes'
require 'ConLivros'
include Fox
include Responder

class FRMVenderLivros < FRMCadastros
  
  ID_FRMVENDERLIVROS_BTNVENDER, ID_FRMVENDERLIVROS_BTNLIMPAR, 
  ID_FRMVENDERLIVROS_BTNFECHAR, ID_FRMVENDERLIVROS_BTNPCLIENTE,
  ID_FRMVENDERLIVROS_BTNPLIVRO, ID_FRMVENDERLIVROS_LSTLIVROS,
  ID_FRMVENDERLIVROS_EDTISBN, ID_FRMVENDERLIVROS_BTNADLIVROS = enum(3000, 8)

  def initialize(clienteMDI, iconeMDI, menuMDI)  
    super(clienteMDI, Resource::STR_VENLIVRO, iconeMDI, menuMDI, 
      TEXTFIELD_NORMAL, false)

    FXMAPFUNC(SEL_COMMAND, ID_FRMVENDERLIVROS_BTNVENDER, :onBtnVender)
    FXMAPFUNC(SEL_COMMAND, ID_FRMVENDERLIVROS_BTNLIMPAR, :onBtnLimpar)
    FXMAPFUNC(SEL_COMMAND, ID_FRMVENDERLIVROS_BTNFECHAR, :onBtnFechar)
    FXMAPFUNC(SEL_CLOSE, 0, :onClose)
    FXMAPFUNC(SEL_COMMAND, ID_FRMVENDERLIVROS_BTNPCLIENTE, :onBtnPCliente)
    FXMAPFUNC(SEL_COMMAND, ID_FRMVENDERLIVROS_BTNPLIVRO, :onBtnPLivro)
    FXMAPFUNC(SEL_COMMAND, ID_FRMVENDERLIVROS_BTNADLIVROS, :onBtnAdLivros)
    FXMAPFUNC(SEL_FOCUSOUT, ID_FRMCADASTROS_EDTCODIGO, :onEdtCodigo_Killfocus)
    FXMAPFUNC(SEL_FOCUSOUT, ID_FRMVENDERLIVROS_EDTISBN, :onEdtISBN_Killfocus)
    FXMAPFUNC(SEL_DOUBLECLICKED, ID_FRMVENDERLIVROS_LSTLIVROS, 
      :onLstLivros_Dblclk)

    resize(500, 370)
    @btnNovo.hide()
  
    frame10 = FXHorizontalFrame.new(@frame3, LAYOUT_LEFT, 0,0,0,0,0,0,0,0)
    frame11 = FXVerticalFrame.new(frame10, LAYOUT_LEFT, 0,0,0,0,0,0,0,0)
  
    frame12 = FXHorizontalFrame.new(@frame3, LAYOUT_LEFT, 0,0,0,0,0,0,0,0)
    frame13 = FXVerticalFrame.new(frame12, LAYOUT_LEFT, 0,0,0,0,0,0,0,0)
    frame14 = FXVerticalFrame.new(frame12, LAYOUT_LEFT, 0,0,0,0,0,0,0,0)
    frame15 = FXVerticalFrame.new(frame12, LAYOUT_LEFT|LAYOUT_BOTTOM, 
      0,0,0,0,0,0,0,0)
  
    frame16 = FXHorizontalFrame.new(@frame3, LAYOUT_LEFT, 0,0,0,0,0,0,0,0)
    frame17 = FXVerticalFrame.new(frame16, LAYOUT_LEFT, 0,0,0,0,0,0,0,0)
    frame18 = FXVerticalFrame.new(frame16, LAYOUT_LEFT, 0,0,0,0,0,0,0,0)
    frame19 = FXHorizontalFrame.new(frame16, LAYOUT_LEFT|LAYOUT_BOTTOM, 
      0,0,0,0,0,0,0,0)
  
    frame20 = FXVerticalFrame.new(@frame3, LAYOUT_LEFT, 0,0,0,0,0,0,0,0)
    frame21 = FXHorizontalFrame.new(frame20, LAYOUT_LEFT, 0,0,0,0,0,0,0,0)
    frame22 = FXHorizontalFrame.new(frame20, LAYOUT_LEFT|FRAME_THICK|
      FRAME_SUNKEN, 0,0,0,0,0,0,0,0)
  
    frame23 = FXHorizontalFrame.new(@frame3, LAYOUT_RIGHT, 0,0,0,0,0,0,0,0)
  
    frame24 = FXHorizontalFrame.new(@frame3, LAYOUT_BOTTOM|LAYOUT_RIGHT)
  
    labDtHrVenda = FXLabel.new(frame11, "Data/Hora Venda:", nil, LAYOUT_LEFT, 
      8, 8, 91, 13)
    @edtDtHrVenda = FXTextField.new(frame11, 30, nil, 0, TEXTFIELD_LIMITED|
      FRAME_SUNKEN|FRAME_THICK|LAYOUT_FIX_WIDTH|LAYOUT_FIX_HEIGHT, 
      8, 24, 128, 21, 3, 20)
    @edtDtHrVenda.disable()
  
    @labCodigo.hide()
    labCodigo = FXLabel.new(frame13, "CPF:", nil, LAYOUT_LEFT, 
      8, 48, 48, 13)
    @edtCodigo.hide()
    @edtCodigo = FXTextField.new(frame13, 14, self, 
      ID_FRMCADASTROS_EDTCODIGO, TEXTFIELD_LIMITED|FRAME_SUNKEN|FRAME_THICK|
      LAYOUT_FIX_WIDTH|LAYOUT_FIX_HEIGHT, 8, 64, 128, 21, 3, 20)
  
    @labDescricao.hide()
    labDescricao = FXLabel.new(frame14, "Cliente:", nil, LAYOUT_LEFT, 
      142, 48, 56, 13)
    @edtDescricao.hide()
    @edtDescricao = FXTextField.new(frame14, 30, nil, 0, TEXTFIELD_LIMITED|
      FRAME_SUNKEN|FRAME_THICK|LAYOUT_FIX_WIDTH|LAYOUT_FIX_HEIGHT, 
      142, 64, 306, 21, 3, 20)
    @edtDescricao.disable()
  
    @btnPCliente = FXButton.new(frame15, "...", nil, self, 
      ID_FRMVENDERLIVROS_BTNPCLIENTE, FRAME_RAISED|FRAME_THICK|
      LAYOUT_FIX_WIDTH|LAYOUT_FIX_HEIGHT|
      LAYOUT_LEFT, 448, 64, 21, 21, 10, 10, 3, 3 )
  
    labISBN = FXLabel.new(frame17, "ISBN:", nil, LAYOUT_LEFT, 
      8, 88, 48, 13)
    @edtISBN = FXTextField.new(frame17, 13, self, 
      ID_FRMVENDERLIVROS_EDTISBN, TEXTFIELD_LIMITED|FRAME_SUNKEN|FRAME_THICK|
      LAYOUT_FIX_WIDTH|LAYOUT_FIX_HEIGHT, 8, 104, 128, 21, 3, 20)
  
    labLivro = FXLabel.new(frame18, "Livro:", nil, LAYOUT_LEFT, 
      142, 88, 40, 13)
    @edtLivro = FXTextField.new(frame18, 30, nil, 0,
      TEXTFIELD_LIMITED|FRAME_SUNKEN|FRAME_THICK|
      LAYOUT_FIX_WIDTH|LAYOUT_FIX_HEIGHT, 142, 104, 282, 21, 3, 20)
    @edtLivro.disable()
  
    @btnAdLivros = FXButton.new(frame19, "+", nil, self, 
      ID_FRMVENDERLIVROS_BTNADLIVROS, FRAME_RAISED|FRAME_THICK|
      LAYOUT_FIX_WIDTH|LAYOUT_FIX_HEIGHT|
      LAYOUT_LEFT, 424, 104, 21, 21, 10, 10, 3, 3 )
    @btnPLivro = FXButton.new(frame19, "...", nil, self, 
      ID_FRMVENDERLIVROS_BTNPLIVRO, FRAME_RAISED|FRAME_THICK|
      LAYOUT_FIX_WIDTH|LAYOUT_FIX_HEIGHT|
      LAYOUT_LEFT, 448, 104, 21, 21, 10, 10, 3, 3 )
  
    labLivros = FXLabel.new(frame21, "Livros para vender:", nil, 
      LAYOUT_LEFT, 8, 128, 120, 13)
    @lstLivros = FXList.new(frame22, 1, self, ID_FRMVENDERLIVROS_LSTLIVROS, 
      LIST_SINGLESELECT|LAYOUT_FIX_WIDTH|LAYOUT_FIX_HEIGHT, 8, 144, 464, 96)
  
    labPrecoTotal = FXLabel.new(frame23, "Preço Total:", nil, 
      LAYOUT_LEFT, 280, 248, 61, 13)
    @edtPrecoTotal = FXTextField.new(frame23, 30, nil, 0,
      TEXTFIELD_LIMITED|FRAME_SUNKEN|FRAME_THICK|
      LAYOUT_FIX_WIDTH|LAYOUT_FIX_HEIGHT, 344, 246, 127, 21, 3, 20)
    @edtPrecoTotal.disable()
  
    @btnVender = FXButton.new(frame24, "&Vender", nil, self, 
      ID_FRMVENDERLIVROS_BTNVENDER, FRAME_RAISED|FRAME_THICK|
      LAYOUT_FIX_WIDTH|LAYOUT_FIX_HEIGHT|
      LAYOUT_LEFT, 239, 275, 75, 25, 15, 15, 3, 3 )
    @btnLimpar = FXButton.new(frame24, "&Limpar", nil, self, 
      ID_FRMVENDERLIVROS_BTNLIMPAR, FRAME_RAISED|FRAME_THICK|
      LAYOUT_FIX_WIDTH|LAYOUT_FIX_HEIGHT|
      LAYOUT_LEFT, 319, 275, 75, 25, 15, 15, 3, 3 )
    @btnFechar = FXButton.new(frame24, "&Fechar", nil, self, 
      ID_FRMVENDERLIVROS_BTNFECHAR, FRAME_RAISED|FRAME_THICK|
      LAYOUT_FIX_WIDTH|LAYOUT_FIX_HEIGHT|
      LAYOUT_LEFT, 399, 275, 75, 25, 15, 15, 3, 3 )
  
    @edtDtHrVenda.setText($Rotinas.retDataHoraStr(true, true))   

    @slPrecos = LISTASTR.new
    @slISBNs = LISTASTR.new
    @slQtdEstoque = LISTASTR.new
    @nPrecoTotal = 0.0
    LimpaDados()
  end

  def LimpaDados()
    @edtDtHrVenda.setText($Rotinas.retDataHoraStr(true, true))  
    @edtCodigo.setText("")
    @edtDescricao.setText("")
    @edtISBN.setText("")
    @edtLivro.setText("")
    @lstLivros.clearItems()
    @nPrecoTotal = 0.0
    @edtPrecoTotal.setText("R$ 0,00")
    @slPrecos.col = []
    @slISBNs.col = []
    @slQtdEstoque.col = []
    @edtCodigo.setFocus()
  end
  
  def ValidaDados()
    if $Rotinas.ValidaCampo(@edtCodigo.getText(), Resource::STR_CPFNINF)
      return false
    end
    if @lstLivros.getNumItems() == 0
      FXMessageBox.error(self, MBOX_OK, "Erro", Resource::STR_LIVVENNINF)
      return false
    end
    return true
  end
  
  def SalvaLista()
    nQtd = @lstLivros.getNumItems()
    sCPF = @edtCodigo.getText()
    sPrecoTotal = @nPrecoTotal.to_s
    
    sDTHR = @edtDtHrVenda.getText()
    sDataHora = $Rotinas.ConverteDataHoraStr(sDTHR, true, 
      "%m/%d/%Y","%H:%M:%S")
    
    for nCont in (0..nQtd-1) do
      sISBN = slISBNs.col[nCont]
    
      $Rotinas.AtualizaDados($Rotinas.gConexao, 
        $ConsSQL.Venda('I',sISBN, sCPF, sDataHora,
        $Rotinas.VirgulaParaPonto(sPrecoTotal,false),""))
    
      nQtdEstoque = @slQtdEstoque.col[nCont].to_i
      if nQtdEstoque > 0
        nQtdEstoque = nQtdEstoque - 1
      end
      sQtdEstoque = nQtdEstoque.to_s
      $Rotinas.AtualizaDados($Rotinas.gConexao, 
        $ConsSQL.Venda('U',sISBN, "","","", sQtdEstoque))
    end
    return true
  end 

  def onBtnVender(sender, sel, ptr)
    if ValidaDados()
      if SalvaLista()
        FXMessageBox.information(self, MBOX_OK, 
          Resource::STR_TITULO, Resource::STR_VENRESUC)
        LimpaDados()
      end
    end
    return true
  end

  def onBtnLimpar(sender, sel, ptr)
    LimpaDados()
    return true
  end

  def onBtnFechar(sender, sel, ptr)
    onClose(sender, sel, ptr)
    return true
  end

  def onClose(sender, sel, ptr)
    $frmPrincipal.habilitaToolButton("tlbVenda")
    $frmPrincipal.habilitaMenu("VenderLivros")
    destroy()
    return true
  end 
 
  def onEdtCodigo_Killfocus(sender, sel, ptr)
    @ConsCampo = LISTAMSTR.new
  
    if @edtCodigo.getText().length > 0
      if $Rotinas.validaCPF(@edtCodigo.getText())
        sCliente = $Rotinas.ConsultaCampoDesc(@ConsCampo, $ConsSQL.Cliente('S',
        @edtCodigo.getText(),"","","","","","","","",""),Resource::STR_CLINENC) 
        if sCliente.length==0
          @edtCodigo.setText("")
          @edtCodigo.killFocus()
          FXMessageBox.error(self, MBOX_OK, "Erro", Resource::STR_CLINENC)
        else
           @edtDescricao.setText(sCliente)
        end
      else
        @edtCodigo.setText("")
        @edtCodigo.killFocus()
        FXMessageBox.error(self, MBOX_OK, "Erro", Resource::STR_CPFINV)
      end
    else
       @edtDescricao.setText("")
    end
    return true
  end
  
  def onEdtISBN_Killfocus(sender, sel, ptr)
    @ConsCampo = LISTAMSTR.new
  
    if @edtISBN.getText().length > 0
      if $Rotinas.validaISBN(@edtISBN.getText())
        sLivro = $Rotinas.ConsultaCampoDesc(@ConsCampo, $ConsSQL.Livro('Q',
        @edtISBN.getText(),"","","","","","","","",""),Resource::STR_LIVNENC)
        if sLivro.length==0
          @edtISBN.setText("")
        	@edtISBN.killFocus()
          FXMessageBox.error(self, MBOX_OK, "Erro", Resource::STR_LIVNENC)
        else
           @edtLivro.setText(sLivro)
           nQtdEstoque = @ConsCampo.col[0][3].to_i
           if nQtdEstoque > 0
             $Rotinas.sPreco = @ConsCampo.col[0][2]
             $Rotinas.sQtdEstoque = @ConsCampo.col[0][3]
           else
             FXMessageBox.error(self, MBOX_OK, "Erro", Resource::STR_LIVNEST)
           end
        end
      else
        @edtISBN.setText("")
        @edtISBN.killFocus()
        FXMessageBox.error(self, MBOX_OK, "Erro", Resource::STR_ISBNINV)
      end
    else
      @edtLivro.setText("")
    end
    return true
  end
  
  def onBtnPCliente(sender, sel, ptr)
    fConClientes = FRMConClientes.new(self)
    fConClientes.execute()
    @edtCodigo.setText($Rotinas.sCodigoSelecionado)
    onEdtCodigo_Killfocus(nil,0,nil)
    return true
  end
  
  def onBtnPLivro(sender, sel, ptr)
    fConLivros = FRMConLivros.new(self)
    fConLivros.execute()
    @edtISBN.setText($Rotinas.sCodigoSelecionado)
    onEdtISBN_Killfocus(nil,0,nil)
    return true
  end
  
  def onBtnAdLivros(sender, sel, ptr) 
    sISBN = @edtISBN.getText()
    sEdtLivro = @edtLivro.getText()
    if sISBN.length > 0
      sLivros = sISBN + " - " + sEdtLivro + " - R$ " + $Rotinas.sPreco
      if @lstLivros.findItem(sLivros)==-1
        @slISBNs.col.push(sISBN)
        @slPrecos.col.push($Rotinas.sPreco)
        @slQtdEstoque.col.push($Rotinas.sQtdEstoque)
        @lstLivros.appendItem(sLivros)
        nPreco = $Rotinas.sPreco.to_f
        @nPrecoTotal = @nPrecoTotal + nPreco
      end
      @edtISBN.setText("")
      @edtLivro.setText("")
      sPrecoTotal = "R$ " + @nPrecoTotal.to_s
      @edtPrecoTotal.setText($Rotinas.VirgulaParaPonto(sPrecoTotal,true))
    end
    return true
  end
  
  def onLstLivros_Dblclk(sender, sel, ptr) 
    if @lstLivros.getNumItems() > 0
      nSelIndice = @lstLivros.getCurrentItem()   
      if @lstLivros.isItemCurrent(nSelIndice)
        nPreco = @slPrecos.col[nSelIndice].to_f
        @nPrecoTotal = @nPrecoTotal-nPreco
        sPrecoTotal = "R$ " + @nPrecoTotal.to_s
        @edtPrecoTotal.setText($Rotinas.VirgulaParaPonto(sPrecoTotal,true))
        @slISBNs.col.delete_at(nSelIndice)
        @slPrecos.col.delete_at(nSelIndice)
        @slQtdEstoque.col.delete_at(nSelIndice)
        @lstLivros.removeItem(nSelIndice)
      end	
    end
    return true
  end  

end
